src/transformers/models/csm/modular_csm.py

```diff
@@ -0,0 +1,1041 @@
+        3 - If labels are provided, expands them to match codebook dimensions and position the target codebook tokens in the inputs_embeds tensor.
+
+        Args:
+            input_ids (`torch.Tensor` of shape `(batch_size, sequence_length)`):
+                The input ids to embed.
+            input_values (`torch.Tensor` of shape `(batch_size, channels, audio_sequence_length)`):
+                The audio input values to embed.
+            input_values_mask (`torch.Tensor` of shape `(batch_size, audio_sequence_length)`):
+                The mask for the input values.
+        """
+        inputs_embeds = self.embed_text_tokens(input_ids)
+
+        if input_values is not None:
+            # =======================================
+            # TODO: @eustlb, this should be batched !!!
+            # but requires making sure batched inference of the codec model works as intended
+            audio_batch_list = [
+                input_values_batch[:, : input_values_mask_batch.sum(-1)]
+                for input_values_batch, input_values_mask_batch in zip(input_values, input_values_mask)
+            ]
+            audio_tokens_list = []
+            for audio_batch in audio_batch_list:
+                codec_outputs = self.codec_model.encode(audio_batch.unsqueeze(0))
+                codebook_ids = codec_outputs.audio_codes.transpose(1, -1)
+                audio_tokens_list.append(codebook_ids[0])
+
+            max_audio_frames = max(el.shape[0] for el in audio_tokens_list)
+            batched_audio_token_ids = torch.stack(
+                [nn.functional.pad(el, (0, 0, 0, max_audio_frames - el.shape[0])) for el in audio_tokens_list]
+            )
+            # =======================================
+
+            audio_embeds = self.backbone_model.embed_tokens(batched_audio_token_ids)
+
+            # batched_audio_codes is a tensor of shape (batch_size, max_audio_frames, num_codebooks)
+            # We need to:
+            # 1. Select only the valid frames for each audio (excluding padding frames)
+            # 2. Embed such frames and place then to their sequence corresponding idxs
+            audio_token_id = self.config.audio_token_id
+            audio_token_mask = input_ids == audio_token_id
+
+            # retreive the number of audio tokens for each audio from the input_ids
+            change_idxs = (audio_token_mask[:, 1:] != audio_token_mask[:, :-1]).nonzero(as_tuple=True)[-1]
+            num_audio_tokens = change_idxs.diff()
+            num_audio_tokens = torch.stack([num_audio_tokens[i] for i in range(0, len(num_audio_tokens), 2)])
+
+            frames_mask = torch.arange(max_audio_frames, device=batched_audio_token_ids.device).expand(
+                len(num_audio_tokens), -1
+            ) < num_audio_tokens.unsqueeze(-1)
+            frames_mask = frames_mask.flatten()
+
+            inputs_embeds[audio_token_mask] = audio_embeds.view(-1, audio_embeds.shape[-1])[frames_mask]
```

Comment:
I maintain that the suggestion is equivalent to what you were doing 😉 https://github.com/huggingface/transformers/pull/36719#discussion_r2067863590